### PR TITLE
Fault Tolerant Manual: Enable the feature

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Find the PoC here: #10568

Parts of #10431

To be merged after #10703

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @justusschock @awaelchli @ninginthecloud

### CODE REVIEW

**Review by ethanwharris — APPROVED** [review]

File: pytorch_lightning/utilities/auto_restart.py [hunk 1]
```diff
@@ -514,7 +512,7 @@ def patch_dataloader_iterator(
 def _add_capture_metadata_collate(dataloader: DataLoader) -> None:
     """Wrap default collate function to retrive captured dataset state dict when fault tolerant is enabled."""
     faut_tolerant_mode = _FaultTolerantMode.detect_current_mode()
-    if not faut_tolerant_mode.is_enabled:
+    if not faut_tolerant_mode.is_enabled or isinstance(dataloader.collate_fn, partial):
```
Comment (awaelchli) [inline pytorch_lightning/utilities/auto_restart.py:515]: is this to check if our collate is already attached? can you remind us in which case can it will occur? also, technically it would be possible that the user has a collate wrapped into partial :)

File: tests/utilities/test_auto_restart.py [hunk 2]
```diff
@@ -1317,3 +1317,163 @@ def test_stateful_workers(num_workers):
...
+        return torch.tensor(index + self.generator.random())
+
+    def state_dict(self):
+        return {self.worker_id: {"random_state": self.generator.getstate()}}
+
+    def load_state_dict(self, state_dict):
+        self._cache_state_dict = state_dict
+
+
+class RandomSamplerStateful(RandomSampler):
```
Comment (awaelchli) [inline tests/utilities/test_auto_restart.py:1353]:
```suggestion
class RandomFaultTolerantSampler(RandomSampler):
```
to follow the naming of the dataset above?

File: tests/utilities/test_auto_restart.py [hunk 3]
```diff
@@ -1317,3 +1317,163 @@ def test_stateful_workers(num_workers):
...
+            self.restarting = False
+
+        for index in indices:
+            self.counter += 1
+            yield index
+
+        self.counter = 0
+
+
+@pytest.mark.skipif(torch.cuda.is_available(), reason="This test takes around 22 sec and should be skipped in Azure CI")
```
Comment (awaelchli) [inline tests/utilities/test_auto_restart.py:1390]: why should we run it on cpu though?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/utilities/auto_restart.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:515]: Yes, this is done to prevent re-wrapping.

I added a better check

```py
(isinstance(dataloader.collate_fn, partial) and dataloader.collate_fn.func is _capture_metadata_collate)
```

File: tests/utilities/test_auto_restart.py (hunk 2 above)
Comment (tchaton) [reply tests/utilities/test_auto_restart.py:1353]: Done !

File: tests/utilities/test_auto_restart.py (hunk 3 above)
Comment (tchaton) [reply tests/utilities/test_auto_restart.py:1390]: Removed. I am not running num_workers as it would take 22 sec.

File: pytorch_lightning/utilities/auto_restart.py [hunk 4]
```diff
@@ -514,12 +512,15 @@ def patch_dataloader_iterator(
 def _add_capture_metadata_collate(dataloader: DataLoader) -> None:
     """Wrap default collate function to retrive captured dataset state dict when fault tolerant is enabled."""
     faut_tolerant_mode = _FaultTolerantMode.detect_current_mode()
-    if not faut_tolerant_mode.is_enabled:
+    collate_fn = dataloader.collate_fn
+    if not faut_tolerant_mode.is_enabled or (
```
Comment (SeanNaren) [inline pytorch_lightning/utilities/auto_restart.py:516]:
```suggestion
    if not fault_tolerant_mode.is_enabled or (
```

File: pytorch_lightning/utilities/auto_restart.py [hunk 5]
```diff
@@ -514,12 +512,15 @@ def patch_dataloader_iterator(
 def _add_capture_metadata_collate(dataloader: DataLoader) -> None:
     """Wrap default collate function to retrive captured dataset state dict when fault tolerant is enabled."""
     faut_tolerant_mode = _FaultTolerantMode.detect_current_mode()
```
Comment (SeanNaren) [inline pytorch_lightning/utilities/auto_restart.py:514]:
```suggestion
    fault_tolerant_mode = _FaultTolerantMode.detect_current_mode()
```

**Review by SeanNaren — APPROVED** [review]

File: pytorch_lightning/utilities/auto_restart.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:516]: Done !

**Review by justusschock — APPROVED** [review]
